File: domain/src/main/java/com/example/domain/usecase/project/core/ValidateInviteCodeUseCase.kt
package com.example.domain.usecase.project.core

import com.example.core_common.result.CustomResult
import com.example.domain.repository.base.ProjectRepository
import javax.inject.Inject

/**
 * 초대 코드를 검증하는 UseCase
 * 
 * @property projectRepository 프로젝트 관련 기능을 제공하는 Repository
 */
class ValidateInviteCodeUseCase @Inject constructor(
    private val projectRepository: ProjectRepository
) {
    /**
     * 초대 코드의 유효성을 검증합니다.
     *
     * @param inviteCode 검증할 초대 코드
     * @return 성공 시 초대 코드 검증 정보, 실패 시 에러 정보가 포함된 Result
     */
    suspend operator fun invoke(inviteCode: String): CustomResult<InviteValidationData, Exception> {
        if (inviteCode.isBlank()) {
            return CustomResult.Failure(Exception("초대 코드가 비어있습니다."))
        }

        return when (val result = projectRepository.validateInviteCode(inviteCode)) {
            is CustomResult.Success -> {
                val valid = result.data["valid"] as? Boolean ?: false
                val projectId = result.data["projectId"] as? String
                val projectName = result.data["projectName"] as? String
                val projectImage = result.data["projectImage"] as? String
                val expiresAt = result.data["expiresAt"] as? Any
                val maxUses = result.data["maxUses"] as? Int
                val currentUses = result.data["currentUses"] as? Int
                val isAlreadyMember = result.data["isAlreadyMember"] as? Boolean ?: false
                val errorMessage = result.data["errorMessage"] as? String

                CustomResult.Success(
                    InviteValidationData(
                        valid = valid,
                        projectId = projectId,
                        projectName = projectName,
                        projectImage = projectImage,
                        expiresAt = expiresAt,
                        maxUses = maxUses,
                        currentUses = currentUses,
                        isAlreadyMember = isAlreadyMember,
                        errorMessage = errorMessage
                    )
                )
            }
            is CustomResult.Failure -> {
                CustomResult.Failure(result.error)
            }
            is CustomResult.Loading -> {
                CustomResult.Loading
            }
            is CustomResult.Initial -> {
                CustomResult.Initial
            }
            is CustomResult.Progress -> {
                CustomResult.Loading
            }
        }
    }
}

/**
 * 초대 코드 검증 데이터 클래스
 */
data class InviteValidationData(
    val valid: Boolean,
    val projectId: String? = null,
    val projectName: String? = null,
    val projectImage: String? = null,
    val expiresAt: Any? = null,
    val maxUses: Int? = null,
    val currentUses: Int? = null,
    val isAlreadyMember: Boolean = false,
    val errorMessage: String? = null
)

File: functions/src/business/member/usecases/validateInviteCode.usecase.ts
import { InviteRepository } from '../../../domain/invite/repositories/invite.repository';
import { ProjectRepository } from '../../../domain/project/repositories/project.repository';
import { MemberRepository } from '../../../domain/member/repositories/member.repository';
import { CustomResult, Result } from '../../../core/types';
import { validateId } from '../../../core/validation';
import { ValidationError } from '../../../core/errors';

export interface ValidateInviteCodeRequest {
  inviteCode: string;
  userId?: string; // Optional - to check if user is already a member
}

export interface ValidateInviteCodeResponse {
  valid: boolean;
  projectId?: string;
  projectName?: string;
  projectImage?: string;
  inviterName?: string;
  expiresAt?: Date;
  maxUses?: number;
  currentUses?: number;
  isAlreadyMember?: boolean;
  errorMessage?: string;
}

export class ValidateInviteCodeUseCase {
  constructor(
    private readonly inviteRepository: InviteRepository,
    private readonly projectRepository: ProjectRepository,
    private readonly memberRepository: MemberRepository
  ) {}

  async execute(request: ValidateInviteCodeRequest): Promise<CustomResult<ValidateInviteCodeResponse>> {
    try {
      const { inviteCode, userId } = request;

      // Validate input
      if (!inviteCode) {
        return Result.failure(
          new ValidationError('request', 'inviteCode is required')
        );
      }

      if (userId) {
        validateId(userId, 'user ID');
      }

      // Find invite by code
      const inviteResult = await this.inviteRepository.findByCode(inviteCode);
      if (!inviteResult.success) {
        return Result.success({
          valid: false,
          errorMessage: 'Failed to validate invite code'
        });
      }

      const invite = inviteResult.data;
      if (!invite) {
        return Result.success({
          valid: false,
          errorMessage: 'Invite code not found'
        });
      }

      // Check if invite is usable
      if (!invite.canBeUsed()) {
        let errorMessage = 'Invite is not usable';
        if (invite.isExpired()) {
          errorMessage = 'Invite has expired';
        } else if (invite.isRevoked()) {
          errorMessage = 'Invite has been revoked';
        } else if (invite.maxUses !== undefined && invite.currentUses >= invite.maxUses) {
          errorMessage = 'Invite has reached maximum usage limit';
        }
        
        return Result.success({
          valid: false,
          errorMessage
        });
      }

      // Get project information
      const projectResult = await this.projectRepository.findById(invite.projectId);
      if (!projectResult.success) {
        return Result.success({
          valid: false,
          errorMessage: 'Failed to validate invite code'
        });
      }

      const project = projectResult.data;
      if (!project) {
        return Result.success({
          valid: false,
          errorMessage: 'Project not found'
        });
      }

      // Check if user is already a member (if userId provided)
      let isAlreadyMember = false;
      if (userId) {
        const existingMemberResult = await this.memberRepository.findByUserId(userId);
        if (existingMemberResult.success && existingMemberResult.data) {
          isAlreadyMember = existingMemberResult.data.isActive();
        }
      }

      return Result.success({
        valid: true,
        projectId: invite.projectId,
        projectName: project.name,
        projectImage: project.image,
        expiresAt: invite.expiresAt,
        maxUses: invite.maxUses,
        currentUses: invite.currentUses,
        isAlreadyMember
      });

    } catch (error) {
      return Result.failure(
        error instanceof Error ? error : new Error('Failed to validate invite code')
      );
    }
  }
}